src/java/htsjdk/samtools/SamReaderFactory.java

```diff
@@ -261,7 +262,7 @@ public SamReader open(final SamInputResource resource) {
                             primitiveSamReader = new BAMFileReader(sourceFile, indexFile, false, validationStringency, this.samRecordFactory);
                         }
                     } else if (BlockCompressedInputStream.isValidFile(bufferedStream)) {
-                        primitiveSamReader = new SAMTextReader(new BlockCompressedInputStream(bufferedStream), validationStringency, this.samRecordFactory);
+                        primitiveSamReader = new SAMTextReader(Defaults.USE_ASYNC_IO_FOR_SAMTOOLS ? new AsyncBlockCompressedInputStream(bufferedStream) : new BlockCompressedInputStream(bufferedStream), validationStringency, this.samRecordFactory);
```

Comment:
the selection of which stream to take is done in the reader class for BAMs - why is it in the reader factory for SAMs? Also, I find it a bit puzzling that it's the stream that is asynchronous and not the reader (on the writing side, we have asynchronous writers and not streams).